### ISSUE TITLE
Adam solver

### ISSUE THREAD

**Issue by zzqiuzz** [jiecaoyu/XNOR-Net-PyTorch#60]
https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/ImageNet/networks/main.py#L326

Hi! Here you set different lr for training,however,for Adam solver, learning rate is automatically changed according to iters. When i set adam solvers in caffe , like 

> net: "train_test.prototxt" 
test_iter: 1000 
test_interval: 1000 
base_lr: 0.001
momentum: 0.9
momentum2: 0.999 
lr_policy: "fixed" 
display: 200 
max_iter: 480000 
weight_decay:0.00005
snapshot: 10000
snapshot_prefix: "snapshot" 
type: "Adam"
solver_mode: GPU

It seems not converge to  that good of yours. Am i wrong in setting hyperparameters in this prototxt?

**Comment by jiecaoyu** [comment]
Hi @zzqiuzz , as shown in [this line](https://github.com/pytorch/pytorch/blob/4edc8273eb223785929c0017caf15c101964d480/torch/optim/adam.py#L105), the Adam solver implemented in PyTorch need to time the gradients with the learning rate to calculate the update steps.

Basically, [Adam](http://ruder.io/optimizing-gradient-descent/index.html#adam) is changing the learning rate relatively for different parameters but it still requires users to decrease the basic learning rate.

A typical learning algorithm which does not require setting the basic learning rate is [Adadelta](http://ruder.io/optimizing-gradient-descent/index.html#adadelta).

**Comment by zzqiuzz** [comment]
@jiecaoyu  Hi! Changing basic learning rates actually improves the performance , lr rates from `0.001` to `1e-6` with in total `56`  training epochs, I get` 45.77% @accuracy1` `69.11%@accuracy5`. Accuracy in top-5 is almostly the same with yours ,however,accuracy in top-1 is little higher than that of yours. I've not updated grads the way in your repo but the way in [torch version](https://github.com/allenai/XNOR-Net/blob/4006d594fe0349a5ed4a5ba6795641d9eecb3aac/util.lua#L78) instead.The model is also derived from that in [torch version.](https://github.com/allenai/XNOR-Net/blob/4006d594fe0349a5ed4a5ba6795641d9eecb3aac/models/alexnetxnor.lua#L32) But a little difference is ,in my implementation, scale layer is added just after bn layer which is the common use in Caffe, while in these works: [HWGQ](https://github.com/zhaoweicai/hwgq/blob/ebc706bee3e2d145de1da4be446ce8de8740738f/examples/imagenet/alex-hwgq-3ne-clip-poly-320k/train_val.prototxt#L75)
 and [TSQ](https://github.com/wps712/Two-Step-Quantization-AlexNet/blob/63d45d7259458af179d7d09d6407321eb1267f32/test_2_ternary.prototxt#L239) they do not add scale layer. so I  wonder if these `scale layers` should be added.

**Comment by jiecaoyu** [comment]
@zzqiuzz Hi, scale is good for improving the accuracy. However, for my implementation, it slows down the training a lot so I didn't add that.
